doc/multicore.md

```diff
@@ -0,0 +1,521 @@
+
+Here's a new version, mixing atomic and non-atomic locations:
+
+```ocaml
+let test run_in_new_domain =
+  let x = ref 5 in
+  let ready = Atomic.make false in
+  Fibre.both
+    (fun () -> run_in_new_domain (fun () ->
+       x := !x * 2;
+       Atomic.set ready true
+    ))
+    (fun () -> run_in_new_domain (fun () ->
+       while not (Atomic.get ready) do () done;
+       traceln "x = %d" !x
+    ))
+```
+
+Initially (before the `Fibre.both`) the memory looks like this:
+
+```
+x     : [ABR]5----------------->
+ready : false   (atomic)
+```
+
+Notice that `ready` now has a frontier too, marked as `R`.
+After writing `x`, we have:
+
+```
+x     : [BR]5--------[A]10----->
+ready : false   (atomic)
+```
+
+When the first branch writes to the atomic location `ready`, `[R]` gets updated too:
+
+```
+x     : [B]5--------[AR]10----->
+ready : true   (atomic)
+```
+
+At this point, if `B` read `x` it might see `5` or `10`, but after reading from `ready` it not only learns that its value is now `true`,
+but also gets its frontier updated with information from `R`:
+
+```
+x     : 5--------[ABR]10----->
+ready : true   (atomic)
+```
+
+Therefore, this version will always print `10`:
+
+```ocaml
+# run test;;
++x = 10
+- : unit = ()
+```
+
+Of course, atomic locations do not store frontiers in a real computer.
+It's just a way of thinking about what values can be read after accessing an atomic location.
+For example, a real system must not perform the write to `ready` before issuing the write to `x`.
+
+### Initialisation
+
+In the above examples, the locations were set up before spawning the new domains.
+What if we create new locations from a domain? For example, what does this do?
+
+```ocaml
+let test run_in_new_domain =
+  let x = ref [1; 2; 3] in
+  Fibre.both
+    (fun () -> run_in_new_domain (fun () ->
+       x := [4; 5; 6]
+    ))
+    (fun () -> run_in_new_domain (fun () ->
+       traceln "x = %a" Fmt.(Dump.list int) !x
+    ))
+```
+
+If the second branch sees the old value of `x` then it will print `[1; 2; 3]`.
+But if it sees the new value (a pointer to a newly allocated list), can we be sure those locations are initialised?
+The answer is yes: a location's timeline in the OCaml memory model starts when the location is allocated
+(it is not a physical memory address which might get reused).
+In particular, an immutable value only ever has one item on its timeline, and so you can only ever read that value.
+So it will always see a correct list:
+
+<!-- $MDX non-deterministic=output -->
+```ocaml
+# run test;;
++x = [1; 2; 3] OR
++x = [4; 5; 6]
+- : unit = ()
+```
+
+The way this works on a real system is interesting.
+Neither `x` nor the list items are atomic, so the system is free to optimise things as it pleases.
+In particular, it might update `x` to point at the new list's address before writing the list.
+However, the new list will be allocated in the writing domain's minor heap, which is private to that domain.
```

Comment:
In the parallel minor GC, all minor heaps can be read by all domains and links can occur between minor heaps. 
To make this work, all mutators must stop to perform minor heap promotion across all domains; we call these periods where all domains are stopped "stop-the-world" sections.